### ISSUE TITLE
Does this version of hnsw works with sparse matrix?

### ISSUE THREAD

**Issue by pommedeterresautee** [nmslib/hnsw#13]
There is no documentation about sparse matrix. 
Does it work with it?

**Comment by yurymalkov** [comment]
Hi @pommedeterresautee,
No, it does not have sparse data support at the moment. I suggest using https://github.com/nmslib/nmslib for search with sparse matrices.
It can be implemented (a new space + update of the space interface + bindings), but probably only with a help of someone who needs such functionality.

**Comment by flying-sheep** [comment]
I would, but writing my thesis I have no time. nmslib sadly has no R bindings though…

**Comment by pommedeterresautee** [comment]
It does!
https://cran.r-project.org/web/packages/RcppHNSW/index.html

**Comment by flying-sheep** [comment]
You linked to RcppHNSW, which wraps hnswlib. RcppNMS doesn’t exist at the time of writing.

**Comment by pommedeterresautee** [comment]
For the full nmslib there is https://cran.r-project.org/web/packages/nmslibR/index.html
But let's be honest it is not really useable. And the api quality is not very high.

**Comment by flying-sheep** [comment]
Oh! Well, I hadn’t found that, thank you! But yeah, a Python-based solution isn’t really helping me much sadly.

**Comment by flying-sheep** [comment]
Yes, thank you! I’ve found that one, but you’re right, Python-based bindings aren’t that helpful.

[Here’s a little overview I made](https://github.com/theislab/destiny/issues/18) about ANN libraries in R

**Comment by searchivarius** [comment]
@flying-sheep what do you mean by "Python bindings not being helpful"? For R, or in general?

Also what does "only sparse" mean in your table in NMSLIB column? You can pass data to NMSLIB using dense matrices, sparse matrices, and strings.

**Comment by flying-sheep** [comment]
> "Python bindings not being helpful"? For R, or in general?

For R! I can’t expect people using my R package to have such a complex setup.

> You can pass data to NMSLIB using dense matrices, sparse matrices, and strings.

Thank you, I missed that in my cursory glance! Fixed
